### PR TITLE
dbr quant: start recording ops which are not quantizeable

### PR DESCRIPTION
**Pull request by vkuzo**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #70372 dbr quant: add limited support for `torch.nn.ModuleList`
* #70349 dbr quant support for custom leaf modules, part 3/x
* #70335 dbr quant: support for custom leaf modules, part 2/x
* #70330 dbr quant: support for custom leaf modules, part 1/x
* #70329 dbr quant: unbreak case when child module not returning any outputs
* #70328 dbr quant: support functional linear without bias
* #70257 dbr quant: support dynamic linear
* #70256 dbr quant: do not attach auto_quant_state to observers
* #70247 dbr quant: stop calling eager quant convert
* #70246 dbr quant: break up test class into multiple classes
* **#70200 dbr quant: start recording ops which are not quantizeable**
* #70196 dbr quant: start reusing static quant module mappings

Summary:

Adds the logic to record not just the subgraphs which are quantizeable,
but also the set of ops (not subgraph) which are quantizeable.  This changes
the information recorded during tracing as follows (an example):

```
// before
1. subgraph of conv1 -> conv2
2. no other information about other ops

// after
1. subgraph of conv1 -> conv2
2. set of types of ops which were not quantizeable but were encountered during tracing
```

This has two uses:
1. easier development of DBR quant to cover more ops, as now the ops which are not being quantized are easier to inspect
2. easier understanding for the user of what DBR quant is doing or not doing for a model

Test plan:

```
python test/test_quantization.py TestQuantizeDBR.test_unsupported_ops_recorded
```

Differential Revision: [D33240997](https://our.internmc.facebook.com/intern/diff/D33240997)

### CODE REVIEW

**Comment by vkuzo** [comment]
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33240997).

**Comment by vkuzo** [comment]
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33240997).

**Comment by vkuzo** [comment]
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33240997).

**Comment by vkuzo** [comment]
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33240997).

File: torch/ao/quantization/_dbr/quantization_state.py [hunk 1]
```diff
@@ -117,6 +117,10 @@ def __init__(
         # to its final value after tracing.
         self.needs_dtype_transform_on_outputs = True
 
+        # For debugging only, stores the types of ops seen by the parent which
+        # did not require op hooks.
+        self.seen_op_types_without_op_hooks: Set[Callable] = set()
```
Comment (jerryzh168) [inline torch/ao/quantization/_dbr/quantization_state.py:122]: so these ops are just added to the set and not connected to the quantizable subgraph right?

**Review by jerryzh168 — APPROVED** [review]

File: torch/ao/quantization/_dbr/quantization_state.py (hunk 1 above)
Comment (vkuzo) [reply torch/ao/quantization/_dbr/quantization_state.py:122]: correct, this is useful when onboarding a new model to check what ops are already supported and what are not.  Once there is parity between DBR and FX graph mode quant, we can delete this code.

**Pull request closed without merging**